File: python/protein-translation/protein_translation.py
CODON_TRANSLATIONS = {
    "AUG": "Methionine",
    "UGG": "Tryptophan",
}
CODON_TRANSLATIONS.update(dict.fromkeys(["UUU", "UUC"], "Phenylalanine"))
CODON_TRANSLATIONS.update(dict.fromkeys(["UUA", "UUG"], "Leucine"))
CODON_TRANSLATIONS.update(dict.fromkeys(["UCU", "UCC", "UCA", "UCG"], "Serine"))
CODON_TRANSLATIONS.update(dict.fromkeys(["UAU", "UAC"], "Tyrosine"))
CODON_TRANSLATIONS.update(dict.fromkeys(["UGU", "UGC"], "Cysteine"))


def is_stop(codon):
    return codon == "UAA" or codon == "UAG" or codon == "UGA"


def proteins(strand):

    if len(strand) % 3 != 0:
        raise ValueError("RNA sequences must only be composed of length-3 codons.")

    proteins_found = []
    while len(strand) > 0:
        codon = strand[0:3]
        strand = strand[3:]

        if is_stop(codon):
            return proteins_found

        proteins_found.append(CODON_TRANSLATIONS[codon])

    return proteins_found

File: javascript/protein-translation/protein-translation.js
const translate_codon = (codon) => {

  switch (codon) {

    case "AUG":	
      return "Methionine";

    case "UUU": case "UUC":	
      return "Phenylalanine";

    case "UUA": case "UUG":
      return "Leucine";

    case "UCU": case "UCC": case "UCA": case "UCG":	
      return "Serine";

    case "UAU": case "UAC":	
      return "Tyrosine";

    case "UGU": case "UGC":	
      return "Cysteine";

    case "UGG":	
      return "Tryptophan";

    default:
      throw new Error('Invalid codon');
  }
}

const is_stop = (codon) => {
  return codon === "UAA" || codon === "UAG" || codon === "UGA";
}

export const translate = (rna_strand) => {

  const codon_pattern = /[A-Z]{3}/g;
  let protein_sequence = []
  let match;

  while (match = codon_pattern.exec(rna_strand)) {

    const codon = match[0];

    if (is_stop(codon)) 
      return protein_sequence;

    protein_sequence.push(translate_codon(codon));
  }
  
  return protein_sequence;
};
